### ISSUE TITLE
whereEq does unnecessary work

### ISSUE THREAD

**Issue by robyoder** [ramda/ramda#2753]
I was looking at [the implementation of `whereEq`](
https://github.com/ramda/ramda/blob/master/source/whereEq.js) and was surprised at what I saw:

```
var whereEq = _curry2(function whereEq(spec, testObj) {
  return where(map(equals, spec), testObj);
});
```

This means if I do something like `filter(whereEq({ type: "a" }), users)`, that test object will be mapped over producing the same final test object for every element in `users`.

Perhaps fixing this is seen as a micro-optimization, and perhaps it is, but it still seems wrong. I’m mobile now, but will run some benchmarks when I get the chance. Basically, though, I’d expect `whereEq(test)` to be equivalent to `where(map(equals, test))`, but it isn’t.

**Comment by robyoder** [comment]
The problem isn't with `where` but with `whereEq`. But yes, in the second thing you linked there was this line:

```
return test ? process(test) : process; // manual currying for performance
```

So that would solve my problem with `whereEq`:

```
var whereEq = function whereEq(spec, testObj) {
  var process = where(map(equals, spec)
  return testObj ? process(testObj) : process; // manual currying for performance
});
```

There may be some other functionality of curry2 that I'm missing, but that's the idea.

**Comment by CrossEye** [comment]
There's a lot of history in all this.  `whereEq` was broken away from `where` at some point because it was such a common case.  But the underlying `where` function had been the cause of some major rewrites.

Very soon after Ramda was publicly announced, a [tweet by the creator of lodash][jd] said the autocurrying was a performance sink.  It was presumably aimed at our new library.  That implementation of `curry` was a major issue, and our fix was to remove the auto-curry function, rewriting with manual currying until we could create a cleaner way to curry.  That performance problem, especially in a library that had been rebuilt to replace an earlier, non-performant, version was a deal-breaker.  With some very important help from early contributor @megawac, we then refactored to include our current internal currying helpers as well as a public `curry` function.  

`where` was a chief function in all this.  It had been prominently featured in [@buzzdecafe's initial announcement][mh] and it was one that helped clearly spell out the differences between Ramda and related libraries.

Now those internal currying helpers also add an additional feature: they manage the use of the placeholder, so that you can write, for instance,

```
const divideBy3 = divide(__, 3)
```

So that forestalls us from doing this sort of manual currying.  I'd like to get past this some day, but I don't see it at the moment.




  [jd]: https://twitter.com/jdalton/status/480141335383400448
  [mh]: https://buzzdecafe.github.io/code/2014/05/16/introducing-ramda

**Comment by robyoder** [comment]
Right, when I looked at it before I remembered that there was something that made it more tricky. The placeholder was it. It shouldn't be too hard to do it manually in this function, but I don't see how to do handle it with the existing curry functions.
